File: src/main/java/cn/jgayb/fenixplugin/MapperUtils.java
package cn.jgayb.fenixplugin;

import com.google.common.base.Function;
import com.google.common.collect.Collections2;
import com.google.common.collect.Lists;
import com.intellij.ide.fileTemplates.FileTemplate;
import com.intellij.ide.fileTemplates.FileTemplateManager;
import com.intellij.ide.fileTemplates.FileTemplateUtil;
import com.intellij.openapi.project.Project;
import com.intellij.psi.*;
import com.intellij.psi.xml.XmlElement;
import com.intellij.util.xml.DomElement;
import com.intellij.util.xml.DomUtil;
import org.jetbrains.annotations.NonNls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.*;


/**
 * @author yanglin
 */
public final class MapperUtils {

    private MapperUtils() {
        throw new UnsupportedOperationException();
    }

    @NotNull
    public static Optional<IdDomElement> findParentIdDomElement(@Nullable PsiElement element) {
        DomElement domElement = DomUtil.getDomElement(element);
        if (null == domElement) {
            return Optional.empty();
        }
        if (domElement instanceof IdDomElement) {
            return Optional.of((IdDomElement) domElement);
        }
        return Optional.ofNullable(DomUtil.getParentOfType(domElement, IdDomElement.class, true));
    }

    public static PsiElement createMapperFromFileTemplate(@NotNull String fileTemplateName,
                                                          @NotNull String fileName,
                                                          @NotNull PsiDirectory directory,
                                                          @Nullable Properties pops) throws Exception {
        FileTemplate fileTemplate = FileTemplateManager.getInstance().getJ2eeTemplate(fileTemplateName);
        return FileTemplateUtil.createFromTemplate(fileTemplate, fileName, pops, directory);
    }

    @NotNull
    public static Collection<PsiDirectory> findMapperDirectories(@NotNull Project project) {
        return Collections2.transform(findMappers(project), new Function<Fenixs, PsiDirectory>() {
            @Override
            public PsiDirectory apply(Fenixs input) {
                return input.getXmlElement().getContainingFile().getContainingDirectory();
            }
        });
    }

    public static boolean isElementWithinFenixFile(@NotNull PsiElement element) {
        PsiFile psiFile = element.getContainingFile();
        return element instanceof XmlElement && DomUtils.isFenixsFile(psiFile);
    }

    @NotNull
    @NonNls
    public static Collection<Fenixs> findMappers(@NotNull Project project) {
        return DomUtils.findDomElements(project, Fenixs.class);
    }

    @NotNull
    @NonNls
    public static Collection<Fenixs> findMappers(@NotNull Project project, @NotNull String namespace) {
        List<Fenixs> result = Lists.newArrayList();
        for (Fenixs fenixs : findMappers(project)) {
            if (getNamespace(fenixs).equals(namespace)) {
                result.add(fenixs);
            }
        }
        return result;
    }

    @NotNull
    public static Collection<Fenixs> findMappers(@NotNull Project project, @NotNull PsiClass clazz) {
        return JavaUtils.isElementWithinInterface(clazz) ? findMappers(project, clazz.getQualifiedName()) : Collections.<Fenixs>emptyList();
    }

    @NotNull
    public static Collection<Fenixs> findMappers(@NotNull Project project, @NotNull PsiMethod method) {
        PsiClass clazz = method.getContainingClass();
        return null == clazz ? Collections.<Fenixs>emptyList() : findMappers(project, clazz);
    }

    @NotNull
    @NonNls
    public static Optional<Fenixs> findFirstMapper(@NotNull Project project, @NotNull String namespace) {
        Collection<Fenixs> fenixs = findMappers(project, namespace);
        return CollectionUtils.isEmpty(fenixs) ? Optional.empty() : Optional.of(fenixs.iterator().next());
    }

    @NotNull
    @NonNls
    public static Optional<Fenixs> findFirstMapper(@NotNull Project project, @NotNull PsiClass clazz) {
        String qualifiedName = clazz.getQualifiedName();
        return null != qualifiedName ? findFirstMapper(project, qualifiedName) : Optional.empty();
    }

    @NotNull
    @NonNls
    public static Optional<Fenixs> findFirstMapper(@NotNull Project project, @NotNull PsiMethod method) {
        PsiClass containingClass = method.getContainingClass();
        return null != containingClass ? findFirstMapper(project, containingClass) : Optional.empty();
    }

    @SuppressWarnings("unchecked")
    @NotNull
    @NonNls
    public static Fenixs getMapper(@NotNull DomElement element) {
        Optional<Fenixs> optional = Optional.ofNullable(DomUtil.getParentOfType(element, Fenixs.class, true));
        if (optional.isPresent()) {
            return optional.get();
        } else {
            throw new IllegalArgumentException("Unknown element");
        }
    }

    @NotNull
    @NonNls
    public static String getNamespace(@NotNull Fenixs fenixs) {
        String ns = fenixs.getNamespace().getStringValue();
        return null == ns ? "" : ns;
    }

    @NotNull
    @NonNls
    public static String getNamespace(@NotNull DomElement element) {
        return getNamespace(getMapper(element));
    }

    @NonNls
    public static boolean isMapperWithSameNamespace(@Nullable Fenixs fenixs, @Nullable Fenixs target) {
        return null != fenixs && null != target && getNamespace(fenixs).equals(getNamespace(target));
    }

    @Nullable
    @NonNls
    public static <T extends IdDomElement> String getId(@NotNull T domElement) {
        return domElement.getId().getRawText();
    }

    @NotNull
    @NonNls
    public static <T extends IdDomElement> String getIdSignature(@NotNull T domElement) {
        return getNamespace(domElement) + "." + getId(domElement);
    }

    @NotNull
    @NonNls
    public static <T extends IdDomElement> String getIdSignature(@NotNull T domElement, @NotNull Fenixs fenixs) {
        Fenixs contextFenixs = getMapper(domElement);
        String id = getId(domElement);
        if (id == null) {
            id = "";
        }
        String idsignature = getIdSignature(domElement);
        //getIdSignature(domElement)
        return isMapperWithSameNamespace(contextFenixs, fenixs) ? id : idsignature;
    }

//    public static void processConfiguredTypeAliases(@NotNull Project project, @NotNull Processor<TypeAlias> processor) {
//        for (Configuration conf : getMybatisConfigurations(project)) {
//            for (TypeAliases tas : conf.getTypeAliases()) {
//                for (TypeAlias ta : tas.getTypeAlias()) {
//                    String stringValue = ta.getAlias().getStringValue();
//                    if (null != stringValue && !processor.process(ta)) {
//                        return;
//                    }
//                }
//            }
//        }
//    }
//
//    private static Collection<Configuration> getMybatisConfigurations(Project project) {
//        return DomUtils.findDomElements(project, Configuration.class);
//    }
//
//    public static void processConfiguredPackage(@NotNull Project project,
//                                                @NotNull Processor<com.wuzhizhan.mybatis.dom.model.Package> processor) {
//        for (Configuration conf : getMybatisConfigurations(project)) {
//            for (TypeAliases tas : conf.getTypeAliases()) {
//                for (com.wuzhizhan.mybatis.dom.model.Package pkg : tas.getPackages()) {
//                    if (!processor.process(pkg)) {
//                        return;
//                    }
//                }
//            }
//        }
//    }
}

File: src/main/java/cn/jgayb/fenixplugin/MapperUtils.kt
package cn.jgayb.fenixplugin

import com.google.common.collect.Collections2
import com.google.common.collect.Lists
import com.intellij.ide.fileTemplates.FileTemplateManager
import com.intellij.ide.fileTemplates.FileTemplateUtil
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiClass
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiMethod
import com.intellij.psi.xml.XmlElement
import com.intellij.util.xml.DomElement
import com.intellij.util.xml.DomUtil
import org.jetbrains.annotations.NonNls
import java.util.*

class MapperUtils private constructor() {
    init {
        throw UnsupportedOperationException()
    }

    companion object {
        @JvmStatic
        fun findParentIdDomElement(element: PsiElement?): Optional<IdDomElement> {
            val domElement = DomUtil.getDomElement(element) ?: return Optional.empty()
            if (domElement is IdDomElement) {
                return Optional.of(domElement)
            }
            return Optional.ofNullable(DomUtil.getParentOfType(domElement, IdDomElement::class.java, true))
        }

        @Throws(Exception::class)
        fun createMapperFromFileTemplate(
            project: Project,
            fileTemplateName: String,
            fileName: String,
            directory: PsiDirectory,
            pops: Properties?
        ): PsiElement {
            val fileTemplate = FileTemplateManager.getInstance(project).getJ2eeTemplate(fileTemplateName)
            return FileTemplateUtil.createFromTemplate(fileTemplate, fileName, pops, directory)
        }

        fun findMapperDirectories(project: Project): Collection<PsiDirectory> {
            return Collections2.transform(
                findMappers(project)
            ) { input -> input!!.xmlElement!!.containingFile.containingDirectory }
        }

        fun isElementWithinFenixFile(element: PsiElement): Boolean {
            val psiFile = element.containingFile
            return element is XmlElement && DomUtils.isFenixsFile(psiFile)
        }

        @NonNls
        fun findMappers(project: Project): Collection<Fenixs> {
            return DomUtils.findDomElements(project, Fenixs::class.java)
        }

        @NonNls
        fun findMappers(project: Project, namespace: String): Collection<Fenixs?> {
            val result: MutableList<Fenixs?> = Lists.newArrayList()
            for (fenixs in findMappers(project)) {
                if (getNamespace(fenixs) == namespace) {
                    result.add(fenixs)
                }
            }
            return result
        }

        fun findMappers(project: Project, clazz: PsiClass): Collection<Fenixs?> {
            return if (JavaUtils.isElementWithinInterface(clazz)) findMappers(
                project,
                clazz.qualifiedName!!
            ) else emptyList<Fenixs>()
        }

        fun findMappers(project: Project, method: PsiMethod): Collection<Fenixs?> {
            val clazz = method.containingClass
            return if (null == clazz) emptyList() else findMappers(project, clazz)
        }

        @NonNls
        fun findFirstMapper(project: Project, namespace: String): Optional<Fenixs> {
            val fenixs = findMappers(project, namespace)
            return if (CollectionUtils.isEmpty(fenixs)) Optional.empty() else Optional.of(
                fenixs.iterator().next()!!
            )
        }

        @NonNls
        fun findFirstMapper(project: Project, clazz: PsiClass): Optional<Fenixs> {
            val qualifiedName = clazz.qualifiedName
            return if (null != qualifiedName) findFirstMapper(project, qualifiedName) else Optional.empty()
        }

        @NonNls
        fun findFirstMapper(project: Project, method: PsiMethod): Optional<Fenixs> {
            val containingClass = method.containingClass
            return if (null != containingClass) findFirstMapper(project, containingClass) else Optional.empty()
        }

        @NonNls
        fun getMapper(element: DomElement): Fenixs {
            val optional = Optional.ofNullable(
                DomUtil.getParentOfType(
                    element,
                    Fenixs::class.java, true
                )
            )
            if (optional.isPresent) {
                return optional.get()
            } else {
                throw IllegalArgumentException("Unknown element")
            }
        }

        @NonNls
        fun getNamespace(fenixs: Fenixs): String {
            val ns = fenixs.getNamespace().stringValue
            return ns ?: ""
        }

        @NonNls
        fun getNamespace(element: DomElement): String {
            return getNamespace(getMapper(element))
        }

        @NonNls
        fun isMapperWithSameNamespace(fenixs: Fenixs?, target: Fenixs?): Boolean {
            return null != fenixs && null != target && getNamespace(fenixs) == getNamespace(
                target
            )
        }

        @NonNls
        fun <T : IdDomElement?> getId(domElement: T): String? {
            return domElement?.id?.rawText
        }

        @NonNls
        fun <T : IdDomElement> getIdSignature(domElement: T): String {
            return getNamespace(domElement) + "." + getId(domElement)
        }

        @NonNls
        fun <T : IdDomElement> getIdSignature(domElement: T, fenixs: Fenixs): String {
            val contextFenixs = getMapper(domElement)
            var id = getId(domElement)
            if (id == null) {
                id = ""
            }
            val idsignature = getIdSignature(domElement)
            return if (isMapperWithSameNamespace(contextFenixs, fenixs)) id else idsignature
        }
    }
}